pkg/serial/handshake.go

```diff
@@ -200,48 +172,56 @@ func HandshakeServer(ctx context.Context, conn net.Conn) error {
 	}
 
 	log.Debug("server: reading syn")
-	// syn is 3 bytes as that will eventually syn us again if we're offset
-	if n, err := readMultiple(conn, syn); n != 2 || err != nil || syn[0] != flagSyn {
-		var msg string
-		if err != nil {
-			msg = fmt.Sprintf("server: failed to read expected SYN: n=%d, err=%s", n, err)
-		} else if syn[0] != flagSyn {
-			msg = fmt.Sprintf("server: did not receive SYN (read %v bytes): %#x != %#x", n, flagSyn, syn[0])
-			conn.Write([]byte{flagNak})
-		} else {
-			msg = fmt.Sprintf("server: received SYN (read %d) bytes", n)
-		}
-
-		// to aid in debug we always dump the full handhsake
-		log.Debug(msg)
-		log.Debugf("server: read %v bytes: ", n)
-		for i := 0; i < n; i++ {
-			log.Debugf("%#x ", syn[i])
-		}
-		log.Debug("")
+	// syn is 2 bytes as that will eventually syn us again if we're offset
+	n, err := io.ReadAtLeast(conn, synBuf, 2)
+	if err != nil {
+		return fmt.Errorf("server: failed to read expected SYN: n=%d, err=%s", n, err)
+	}
 
-		return errors.New(msg)
+	if synBuf[0] != flagSyn {
+		conn.Write([]byte{flagNak})
+		// to aid in debug we always dump the full handshake
+		log.Debugf("server: read %v bytes: %#v", n, synBuf[:n])
+		return fmt.Errorf("server: did not receive SYN (read %v bytes): %#x != %#x", n, flagSyn, synBuf[0])
 	}
-	log.Debugf("server: received syn: %#x\n", syn)
+
+	log.Debugf("server: received syn: %#v", synBuf)
 
 	log.Debug("server: writing synack")
-	synack[0] = flagAck
-	synack[1] = syn[1] + 1
-	rand.Read(synack[2:])
 
-	conn.Write(synack)
+	// last byte is random.
+	synackBuf := []byte{flagAck, synBuf[1] + 1, 0}
+	rand.Read(synackBuf[2:])
+
+	conn.Write(synackBuf)
 
-	ack[0] = flagAck
-	ack[1] = synack[2] + 1
+	ack := []byte{flagAck, synackBuf[2] + 1}
 	log.Debug("server: reading ack")
-	readMultiple(conn, buf)
-	if (buf[0] != flagAck && buf[0] != flagDebugAck) || bytes.Compare(ack[1:], buf[1:]) != 0 {
-		msg := fmt.Sprintf("server: did not receive ack: %#x != %#x", ack, buf)
-		log.Debug(msg)
+
+	buf := make([]byte, 2)
+	io.ReadFull(conn, buf)
```

Comment:
This will not rapidly drain spurious data on the line as you're only reading 2 bytes max. You'll have to loop repeatedly to do so.